lib/compat/wordpress-6.1/script-loader.php

```diff
@@ -36,6 +40,21 @@ static function () use ( $style ) {
 	);
 }
 
+/**
+ * Fetches, processes and compiles stored block supports styles, then renders them to the page.
+ * Styles are stored via the style engine API. See: packages/style-engine/README.md
+ */
+function gutenberg_enqueue_stored_block_supports_styles() {
+	$styles  = gutenberg_style_engine_get_stylesheet_from_store( 'block-supports' );
+	$styles .= gutenberg_style_engine_get_stylesheet_from_store( 'layout-block-supports' );
+	if ( ! empty( $styles ) ) {
+		$key = 'block-supports-styles';
+		wp_register_style( $key, false, array(), true, true );
+		wp_add_inline_style( $key, $styles );
+		wp_enqueue_style( $key );
+	}
+}
+
```

Comment:
I think it would be better if we abstract this function a bit... Instead of having a function specifically for the block-supports, we could have it for all stores. In another PR I'm converting global-styles to use the style-engine, so if we make this function specifically for block-supports, then we'd have to add a separate function for global-styles, then another one for theme-styles and any other implementation we may come up with in the future.

This works fine:
```suggestion
/**
 * Fetches, processes and compiles stored styles, then renders them to the page.
 * Styles are stored via the style engine API. See: packages/style-engine/README.md
 */
function gutenberg_enqueue_stored_styles() {
	$stores = WP_Style_Engine_CSS_Rules_Store_Gutenberg::get_stores();
	foreach ( array_keys( $stores ) as $store_name ) {
		$styles = gutenberg_style_engine_get_stylesheet_from_store( $store_name );
		if ( ! empty( $styles ) ) {
			$key = "sp-style-engine-$store_name";
			wp_register_style( $key, false, array(), true, true );
			wp_add_inline_style( $key, $styles );
			wp_enqueue_style( $key );
		}
	}
}

```

Reply:
Fair call on the name change. Much better. 

I was also wondering about how we'd cater for global styles and make it backwards compatible.

I left it out of the equation so we could decide later how to handle them. Mainly because of what's going on in `gutenberg_enqueue_global_styles()` and the ever-changing nature of global styles.  

Also, this would print out every stored store. Is that the intention? 

Would there be any issue against being explicit about which styles Gutenberg prints, i.e., adding them as we need them?

Aside from the certainty/control aspect, we'll need to determine the order eventually, e.g., we definitely want block supports styles to be printed after global styles.

I can imagine it would be handy for theme authors to be able to queue styles via the style engine, but I'm unaware if it's something that they should control themselves (?)

Reply:
Oh, great discussion here, I think I agree with both of you 😄 — it'd be good to come up with a combined function that works for those future use cases, rather than needing it to be tied specifically to block supports 👍

> Would there be any issue against being explicit about which styles Gutenberg prints, i.e., adding them as we need them?

I agree that it'd be better to be explicit about which particular stores we're enqueueing, so that we can both determine the order, and have the function be clear about what's being enqueued (which should help with debugging).

Ideally, if something somewhere (a plugin or theme) adds another store, then maybe it shouldn't be automatically enqueued. Enqueueing styles should probably still be the responsibility of the theme to do explicitly, rather than happening automatically?

If we want the function to help themes or other use cases with the task of enqueueing, perhaps one way to do it might be to have a param that defaults to `default` or something like that. If it's `default` then we enqueue the Gutenberg stores, and if it isn't `default` then it looks up the store name to be enqueued? 🤷

Also, this might just be a personal preference, but it can be useful to have explicit strings for style keys where we can, so that if you're inspecting a rendered page and need to look up where that style gets generated, you can search the codebase for the string in order to find it. Not very DRY, but can sometimes help with debugging.

Reply:
> it'd be good to come up with a combined function that works for those future use cases, rather than needing it to be tied specifically to block supports

Yeah, it'd be great to be able satisfy both causes:

- having control over the stores that Gutenberg requires to be enqueued
- giving theme authors the opportunity to opt-in to this behaviour

A naive version might be to filter out and enqueue Gutenberg's stores in `gutenberg_enqueue_block_support_styles()`, then loop over the remaining (presumably added elsewhere by themes or whatever).

It'd be better to explicitly define "equeuability" so maybe we could utilize the `context` and `enqueue` parameters so that we store all styles that specify a context, but only enqueue them if `enqueue === true` 🤔 

The consequence is that we'd have to filter out rules from the store that shouldn't be enqueued. 🤔 🤔🤔🤔🤔🤔

```php
return gutenberg_style_engine_get_stylesheet(
    $layout_styles,
   array(
      'context' => 'layout-block-supports', // It will be stored in the  'layout-block-supports' store.
      'enqueue' => true, // It will be enqueued.
   )
);
```

Reply:
> A naive version might be to filter out and enqueue Gutenberg's stores in gutenberg_enqueue_block_support_styles(), then loop over the remaining (presumably added elsewhere by themes or whatever).



<details>

<summary>I mean something like this:</summary>

```php
/**
 * Fetches, processes and compiles stored core styles, then combines and renders them to the page.
 * Any other registered stores are enqueued separately after the core styles.
 * Styles are stored via the style engine API. See: packages/style-engine/README.md
 */
function gutenberg_enqueue_stored_styles() {
	$core_styles_keys    = array( 'block-supports', 'layout-block-supports' );
	$compiled_stylesheet = '';
	foreach ( $core_styles_keys as $style_key ) {
		$compiled_stylesheet .= gutenberg_style_engine_get_stylesheet_from_store( $style_key );
	}

	// Combine Core styles.
	// @TODO check for `defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG` and split them for ease of debugging.
	if ( ! empty( $styles ) ) {
		$key = 'block-supports-styles';
		wp_register_style( $key, false, array(), true, true );
		wp_add_inline_style( $key, $compiled_stylesheet );
		wp_enqueue_style( $key );
	}

	$additional_stores = WP_Style_Engine_CSS_Rules_Store_Gutenberg::get_stores();
	foreach ( array_keys( $additional_stores ) as $store_name ) {
		if ( in_array( $store_name, $core_styles_keys, true ) ) {
			continue;
		}
		$styles = gutenberg_style_engine_get_stylesheet_from_store( $store_name );
		if ( ! empty( $styles ) ) {
			$key = "wp-style-engine-$store_name";
			wp_register_style( $key, false, array(), true, true );
			wp_add_inline_style( $key, $styles );
			wp_enqueue_style( $key );
		}
	}
}

```

</details>

Reply:
> Also, this would print out every stored store. Is that the intention?

Yes

> Would there be any issue against being explicit about which styles Gutenberg prints, i.e., adding them as we need them?

No issue with that, but I don't think that having a separate function per-store would be the right way to approach this 😅 

> Aside from the certainty/control aspect, we'll need to determine the order eventually, e.g., we definitely want block supports styles to be printed after global styles.

Well... not necessarily! I agree with the _"we'll need to determine the order eventually"_ part of the sentence, but I don't believe they should actually be enqueued separately!
Ideally, block supports and global styles would be in the same store (let's say a `core` store).
Block supports would be added to the store _after_ global styles, so any CSS that needs to override global-styles will be merged and override things properly. Having these 2 in the same store would allow for further optimization of the resulting CSS by combining things where appropriate 😉 

Maybe instead of a priority on enqueue it would be more appropriate to have a priority at the time we add styles to the store... This way we'd be able to sort the array based on priorities, and merge styles properly, ensuring overrides are correctly applied?

> I can imagine it would be handy for theme authors to be able to queue styles via the style engine, but I'm unaware if it's something that they should control themselves (?)

They definitely should be able to! I can think of many scenarios where theme-authors would be able to take advantage of it and append their styles to the Core styles, or use the style-engine to add block-specific styles! It would certainly be a lot more efficient (for performance and sustainability) if they added their opinionated block-styles this way instead of enqueueing monolithic stylesheets.
Additionally, it's not just themes that can use it... Think `block.json` and styles we can add in there - both in Core blocks and 3rd-party blocks.

> Ideally, if something somewhere (a plugin or theme) adds another store, then maybe it shouldn't be automatically enqueued. Enqueueing styles should probably still be the responsibility of the theme to do explicitly, rather than happening automatically?

Hmmm what if we add a prop to stores that can define if the store should be automatically enqueued or not?

> Also, this might just be a personal preference, but it can be useful to have explicit strings for style keys where we can, so that if you're inspecting a rendered page and need to look up where that style gets generated, you can search the codebase for the string in order to find it. Not very DRY, but can sometimes help with debugging.

Intersting. I suppose we could do something like that if `SCRIPT_DEBUG` is set to `true`... but I don't think it should be the default behavior. The default output should always be optimized for consumers, not developers 😉

Reply:
> A naive version might be to filter out and enqueue Gutenberg's stores in gutenberg_enqueue_block_support_styles(), then loop over the remaining (presumably added elsewhere by themes or whatever).

Seems reasonable!

Reply:
> No issue with that, but I don't think that having a separate function per-store would be the right way to approach this

Totally agree 👍 

> I agree with the "we'll need to determine the order eventually" part of the sentence, but I don't believe they should actually be enqueued separately!

I think we're on the same page. I was only referring to having the ability, however we do it, to respect the styles hierarchy and preserve the cascade. 

So I mean `Global styles -> Elements -> Block styles -> Block supports`. 

> Ideally, block supports and global styles would be in the same store (let's say a core store).

This makes sense! I like!

> Block supports would be added to the store after global styles, so any CSS that needs to override global-styles will be merged and override things properly. Having these 2 in the same store would allow for further optimization of the resulting CSS by combining things where appropriate

Oooh now I see what you mean. Sweet! Thanks for the explainer. 🍺 

I'm off for a couple of days but will finish this off and do some more testing on Friday. 

Thanks for all the great feedback!

Reply:
> A naive version might be to filter out and enqueue Gutenberg's stores in gutenberg_enqueue_block_support_styles(), then loop over the remaining (presumably added elsewhere by themes or whatever).

I've done this for now and have added test coverage for it. 

Happy to refine things in follow ups!